File: insertion-sort/insertion-sort.cs
using System;
using System.Diagnostics;

public class InsertionSortClass
{
    static int GenerateIndex(int num, Random rand) => rand.Next(num);

    static void InsertIntoArray(int[] arr, int pos, int elem)
    {
        for (var i = arr.Length - 1; i > pos; i--) {
            arr[i] = arr[i - 1];
        }
        arr[pos] = elem;
    }

    static int[] GenerateArray(int size)
    {
        var rand = new Random();
        var arr = new int[size];
        for (var i = 1; i <= size; i++) {
            var index = GenerateIndex(i, rand);
            InsertIntoArray(arr, index, i);
        }
        return arr;
    }

    static void InsertionSort(int[] arr)
    {
        for (var i = 1; i < arr.Length; i++) {
            var key = arr[i];
            var j = i - 1;
            while (j >= 0 && arr[j] > key) {
                arr[j + 1] = arr[j];
                j--;
            }
            arr[j + 1] = key;
        }
    }

    static int[] GetExpectedArray(int size)
    {
        var arr = new int[size];
        for (var i = 0; i < size; i++) arr[i] = i + 1;
        return arr;
    }

    static bool CompareArrays(int[] arr, int[] expected)
    {
        for (var i = 0; i < arr.Length; i++) {
            if (arr[i] != expected[i]) return false;
        }
        return true;
    }

    public static void Main(string[] args)
    {
        var size = 30000;

        // Generate array
        var sw1 = new Stopwatch();
        sw1.Start();
        var arr = GenerateArray(size);
        sw1.Stop();

        Console.WriteLine("1. Time to generate array: " + sw1.Elapsed.Milliseconds);

        // Sort array
        var sw2 = new Stopwatch();
        sw2.Start();
        InsertionSort(arr);
        sw2.Stop();

        // Eval sorted array with an expected result
        var expected = GetExpectedArray(size);
        bool are_equal = CompareArrays(arr, expected);
        if (! are_equal) {
            Console.WriteLine("2. The array is NOT sorted as expected");
        } else {
            Console.WriteLine("2. Result array is sorted as expected");
            Console.WriteLine("3. Time to sort the array: " + sw2.Elapsed.Milliseconds);
        }
    }
}

File: insertion-sort/insertion-sort.ex
defmodule InsertionSort do
    defp generate_array(size) do
        Enum.reduce(1..size, [], fn i, acc ->
            index = Enum.random(0..i)
            List.insert_at(acc, index, i)
        end)
    end

    defp get_index([], _key, index), do: index

    defp get_index(list, key, index) do
        [head | body] = list
        if head >= key do
            index
        else
            get_index(body, key, index + 1)
        end
    end

    defp insertion_sort(arr) do
        Enum.reduce(arr, [], fn key, acc ->
            index = get_index(acc, key, 0)
            List.insert_at(acc, index, key)
        end)
    end

    defp compare_lists(subject, expected) do
        n_diff = Enum.reduce(0..length(expected) - 1, 0, fn i, acc ->
            if Enum.at(subject, i) == Enum.at(expected, i) do
                acc
            else
                acc + 1
            end
        end)
        n_diff == 0
    end

    def main() do # 4653 ms
        size = 30_000

        # Generate array
        {gen_time, arr} = :timer.tc(fn -> generate_array(size) end)
        # arr |> Kernel.inspect() |> IO.puts()

        # Sort array
        {sort_time, res} = :timer.tc(fn -> insertion_sort(arr) end)
        # res |> Kernel.inspect() |> IO.puts()

        # Eval sorted array with expected result
        expected = Enum.to_list(1..size)
        are_equal = compare_lists(res, expected)

        # Output results
        IO.puts "1. Time to generate list: #{gen_time / 1_000} ms"
        if not are_equal do
            IO.puts "2. The result list is NOT sorted as expected"
        else
            IO.puts "2. The result list is sorted as expected"
            IO.puts "3. Time to sort the list #{sort_time / 1_000} ms"
        end
    end
end

InsertionSort.main()
